File: cogs/moderation/moderation_cog.py
import discord
from discord.ext import commands
from utils.error_handler import error_handler
from utils.permissions import has_permissions_to_send_messages

class ModerationCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    async def _kick_or_ban(self, ctx, users, action, reason):
        action_func = getattr(users[0], action)
        affected_users = []
        for user in users:
            await action_func(reason=reason)
            affected_users.append(user.name)
        await ctx.send(f'{len(affected_users)} users have been {action}ed from the server. Reason: {reason if reason else "No reason provided"}.')

    @commands.command(name='moderate')
    @error_handler
    @has_permissions_to_send_messages()
    async def moderate(self, ctx, action: str, user: discord.Member, *, reason=None):
        if action.lower() not in ['kick', 'ban']:
            await ctx.send("Invalid action. Please choose 'kick' or 'ban'.")
            return
        await self._kick_or_ban(ctx, [user], action, reason)

    @commands.command(name='massmoderate')
    @error_handler
    @has_permissions_to_send_messages()
    async def mass_moderate(self, ctx, action: str, users: commands.Greedy[discord.Member], *, reason=None):
        if action.lower() not in ['kick', 'ban']:
            await ctx.send("Invalid action. Please choose 'kick' or 'ban'.")
            return
        await self._kick_or_ban(ctx, users, action, reason)

    @commands.command(name='moderaterole')
    @error_handler
    @has_permissions_to_send_messages()
    async def moderate_role(self, ctx, action: str, role: discord.Role, *, reason=None):
        if action.lower() not in ['kick', 'ban']:
            await ctx.send("Invalid action. Please choose 'kick' or 'ban'.")
            return
        members = [member for member in ctx.guild.members if role in member.roles]
        await self._kick_or_ban(ctx, members, action, reason)

def setup(bot):
    bot.add_cog(ModerationCog(bot))

File: cogs/moderation/moderation_cog.ts
import { Client, Message, GuildMember, Role } from 'discord.js-selfbot-v13';
import { Command } from 'discord-akairo';
import { errorHandler } from '../../utils/error_handler';
import { hasPermissionsToSendMessages } from '../../utils/permissions';
import { kickOrBan } from '../../utils/moderation_utils';

class ModerationCog {
    private client: Client;

    constructor(client: Client) {
        this.client = client;
    }

    @Command({
        aliases: ['moderate'],
        description: 'Moderate a user by kicking or banning them from the server.',
        args: [
            { id: 'action', type: 'string' },
            { id: 'user', type: 'member' },
            { id: 'reason', type: 'string', match: 'rest', default: null }
        ]
    })
    @errorHandler
    @hasPermissionsToSendMessages()
    async moderate(message: Message, { action, user, reason }: { action: string, user: GuildMember, reason: string | null }) {
        if (!['kick', 'ban'].includes(action.toLowerCase())) {
            await message.channel.send("Invalid action. Please choose 'kick' or 'ban'.");
            return;
        }
        if (!user) {
            await message.channel.send("Invalid user. Please mention a valid user.");
            return;
        }
        await kickOrBan(message, [user], action as 'kick' | 'ban', reason);
    }

    @Command({
        aliases: ['massmoderate'],
        description: 'Moderate multiple users by kicking or banning them from the server.',
        args: [
            { id: 'action', type: 'string' },
            { id: 'users', type: 'members' },
            { id: 'reason', type: 'string', match: 'rest', default: null }
        ]
    })
    @errorHandler
    @hasPermissionsToSendMessages()
    async massModerate(message: Message, { action, users, reason }: { action: string, users: GuildMember[], reason: string | null }) {
        if (!['kick', 'ban'].includes(action.toLowerCase())) {
            await message.channel.send("Invalid action. Please choose 'kick' or 'ban'.");
            return;
        }
        await kickOrBan(message, users, action as 'kick' | 'ban', reason);
    }

    @Command({
        aliases: ['moderaterole'],
        description: 'Moderate all users with a specific role by kicking or banning them from the server.',
        args: [
            { id: 'action', type: 'string' },
            { id: 'role', type: 'role' },
            { id: 'reason', type: 'string', match: 'rest', default: null }
        ]
    })
    @errorHandler
    @hasPermissionsToSendMessages()
    async moderateRole(message: Message, { action, role, reason }: { action: string, role: Role, reason: string | null }) {
        if (!['kick', 'ban'].includes(action.toLowerCase())) {
            await message.channel.send("Invalid action. Please choose 'kick' or 'ban'.");
            return;
        }
        const members = role.members.array();
        await kickOrBan(message, members, action as 'kick' | 'ban', reason);
    }
}

export function setup(client: Client) {
    return new ModerationCog(client);
}
